tiledb/api/src/query/read/output/strategy.rs

```diff
@@ -221,12 +221,12 @@ where
                             .into_iter()
                             .take(ncells + 1)
                             .collect::<Vec<u64>>();
-                        if let Some(first) = offsets.first_mut() {
-                            *first = 0u64;
-                        }
                         if let Some(last) = offsets.last_mut() {
                             *last = nvalues as u64;
                         }
+                        if let Some(first) = offsets.first_mut() {
+                            *first = 0u64;
+                        }
```

Comment:
This is just out of curiosity, but am I missing a reason for this or is it just a "temporary thing that got refactored out" change?

Reply:
This one actually was functional!

(side note: I think it would be wise to also have the write side do the extra offsets, for consistency, plus we can likely get rid of this sort of thing)

Because the write side doesn't have the extra offset, if you have one cell, then you have one offset.  If you have one offset, then `first_mut()` and `last_mut()` return the same memory location.  The first offset is always supposed to be zero, but it wouldn't always be so without swapping the order!

Reply:
The number of values of cell `i` is determined by `offsets[i + 1] - offsets[i]` and are element-based.  And then `offsets[i]` is the index of the first value for cell `i`.

```
                proptest::collection::vec(
                    0u64..=(nvalues as u64),
                    0..=max_offsets,
                ),
```
This part of the strategy picks a random set of indices into the values.

```
                        offsets = offsets
                            .into_iter()
                            .take(ncells + 1)
                            .collect::<Vec<u64>>();
                        if let Some(last) = offsets.last_mut() {
                            *last = nvalues as u64;
                        }
                        if let Some(first) = offsets.first_mut() {
                            *first = 0u64;
                        }
                        offsets.sort();
```
This part of the strategy makes sure that the set of random indices conforms to the other required properties, which is that (1) the offsets begin at zero; (2) the offsets end with the number of values; and (3) the offsets are sorted.

If we have 10 values then maybe the initial part gives us `[4, 6, 3, 7, 1, 1, 3]`, which would turn into `[0, 6, 3, 7, 1, 1, 10]` before sorting, and then `[0, 1, 1, 3, 6, 7, 10]` after sorting.

Reply:
I told you my brain wasn't working last night. I was totally missing the `offsets.sort()` step. Nice!